File: datamodel/src/main/kotlin/de/westermann/robots/datamodel/util/Energy.kt
package de.westermann.robots.datamodel.util

/**
 * @author lars
 */
data class Energy(
        val value: Double,
        val state: State
) {
    companion object {
        val UNKNOWN = Energy(0.0, State.UNKNOWN)

        fun parse(data: String) = data.split(",").let {
            Energy(it[0].toDouble(), State.valueOf(it[1]))
        }

        fun fromJson(json: Json) = Energy(
                json["value"]?.toString()?.toDoubleOrNull() ?: 0.0,
                json["state"]?.toString()?.let { s ->
                    State.values().find { it.name.equals(s, ignoreCase = true) }
                } ?: State.UNKNOWN
        )
    }

    fun toMqtt() = "$value,$state"

    fun toJson() = json {
        value("value") { value }
        value("state") { state.name }
    }


    enum class State {
        CHARGING, DISCHARGING, UNKNOWN, NO_BATTERY
    }
}

File: python/src/util/energy.py
from enum import Enum


class EnergyState(Enum):
    CHARGING = 0
    DISCHARGING = 1
    UNKNOWN = 2
    NO_BATTERY = 3


class Energy:
    def __init__(self, value: float, state: EnergyState):
        self.value = value
        self.state = state

    def __str__(self):
        return str(self.value) + "," + str(self.state.name)

    def __eq__(self, other):
        if (not isinstance(other, Energy)) or (other is None):
            return False

        return self.value == other.value and self.state == other.state

    def __ne__(self, other):
        return not self == other

    @staticmethod
    def parse(value: str) -> "Energy":
        h = value.split(",")
        return Energy(float(h[0]), EnergyState[h[1]])
